test/hotspot/jtreg/testlibrary/ctw/src/sun/hotspot/tools/ctw/CtwRunner.java

```diff
@@ -258,43 +267,54 @@ private String[] cmd(long classStart, long classStop) {
         String phase = phaseName(classStart);
         Path file = Paths.get(phase + ".cmd");
         var rng = Utils.getRandomInstance();
+
+        ArrayList<String> Args = new ArrayList<String>(Arrays.asList(
+                "-Xbatch",
+                "-XX:-UseCounterDecay",
+                "-XX:-ShowMessageBoxOnError",
+                "-XX:+UnlockDiagnosticVMOptions",
+                // redirect VM output to cerr so it won't collide w/ ctw output
+                "-XX:+DisplayVMOutputToStderr",
+                // define phase start
+                "-DCompileTheWorldStartAt=" + classStart,
+                "-DCompileTheWorldStopAt=" + classStop,
+                // CTW library uses WhiteBox API
+                "-XX:+WhiteBoxAPI", "-Xbootclasspath/a:wb.jar",
+                // export jdk.internal packages used by CTW library
+                "--add-exports", "java.base/jdk.internal.jimage=ALL-UNNAMED",
+                "--add-exports", "java.base/jdk.internal.misc=ALL-UNNAMED",
+                "--add-exports", "java.base/jdk.internal.reflect=ALL-UNNAMED",
+                "--add-exports", "java.base/jdk.internal.access=ALL-UNNAMED",
+                // enable diagnostic logging
+                "-XX:+LogCompilation",
+                // use phase specific log, hs_err and ciReplay files
+                String.format("-XX:LogFile=hotspot_%s_%%p.log", phase),
+                String.format("-XX:ErrorFile=hs_err_%s_%%p.log", phase),
+                String.format("-XX:ReplayDataFile=replay_%s_%%p.log", phase),
+                // MethodHandle MUST NOT be compiled
+                "-XX:CompileCommand=exclude,java/lang/invoke/MethodHandle.*",
+                // Stress* are c2-specific stress flags, so IgnoreUnrecognizedVMOptions is needed
+                "-XX:+IgnoreUnrecognizedVMOptions",
+                "-XX:+StressLCM",
+                "-XX:+StressGCM",
+                "-XX:+StressIGVN",
+                "-XX:+StressCCP",
+                // StressSeed is uint
+                "-XX:StressSeed=" + Math.abs(rng.nextInt())));
+
+        System.out.println(CTW_EXTRA_ARGS);
+        if (null != CTW_EXTRA_ARGS) {
+            for (String arg : CTW_EXTRA_ARGS.split(",")) {
```

Comment:
I read CtwRunner.java. It is capable of restarting the failed JVM process after shifting the offset. All we need are a couple of tweaks. 

1. workaround wb.jar. 
2. add a way to pass JVM options from outside. 

that's my understanding of this PR. right?

Reply:
yup